sdk/helper/random/parser.go

```diff
@@ -0,0 +1,250 @@
+package random
+
+import (
+	"fmt"
+	"reflect"
+	"sort"
+	"unicode"
+	"unicode/utf8"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/hcl"
+	"github.com/mitchellh/mapstructure"
+)
+
+// Parse is a convenience function for parsing HCL into a StringGenerator. See Parser.Parse for details.
+func Parse(raw string) (strs StringGenerator, err error) {
+	parser := Parser{
+		RuleRegistry: Registry{
+			Rules: defaultRuleNameMapping,
+		},
+	}
+	return parser.Parse(raw)
+}
+
+// Parser parses string generator configuration from HCL.
+type Parser struct {
+	// RuleRegistry maps rule names in HCL to Rule constructors.
+	RuleRegistry Registry
+}
+
+// Parse parses the provided HCL into a StringGenerator.
+func (p Parser) Parse(raw string) (strs StringGenerator, err error) {
+	rawData := map[string]interface{}{}
+	err = hcl.Decode(&rawData, raw)
+	if err != nil {
+		return strs, fmt.Errorf("unable to decode: %w", err)
+	}
+
+	// Decode the top level items
+	decoder, err := mapstructure.NewDecoder(&mapstructure.DecoderConfig{
+		Result:     &strs,
+		DecodeHook: stringToRunesFunc,
+	})
+	if err != nil {
+		return strs, fmt.Errorf("unable to decode configuration: %w", err)
+	}
+
+	err = decoder.Decode(rawData)
+	if err != nil {
+		return strs, fmt.Errorf("failed to decode configuration: %w", err)
+	}
+
+	// Decode & parse rules
+	rawRules, err := getMapSlice(rawData, "rule")
+	if err != nil {
+		return strs, fmt.Errorf("unable to retrieve rules: %w", err)
+	}
+
+	rules, err := p.parseRules(rawRules)
+	if err != nil {
+		return strs, fmt.Errorf("unable to parse rules: %w", err)
+	}
+
+	// Add any charsets found in rules to the overall charset & deduplicate
+	cs := append(strs.Charset, getChars(rules)...)
+	cs = deduplicateRunes(cs)
+
+	strs = StringGenerator{
+		Length:  strs.Length,
+		Charset: cs,
+		Rules:   rules,
+	}
+
+	err = validate(strs)
+	if err != nil {
+		return strs, err
+	}
+
+	return strs, nil
+}
+
+func (p Parser) parseRules(rawRules []map[string]interface{}) (rules []Rule, err error) {
+	for _, rawRule := range rawRules {
+		info, err := getRuleInfo(rawRule)
+		if err != nil {
+			return nil, fmt.Errorf("unable to get rule info: %w", err)
+		}
+
+		// Map names like "lower-alpha" to lowercase alphabetical characters
+		applyShortcuts(info.data)
+
+		rule, err := p.RuleRegistry.parseRule(info.ruleType, info.data)
+		if err != nil {
+			return nil, fmt.Errorf("unable to parse rule %s: %w", info.ruleType, err)
+		}
+		rules = append(rules, rule)
+	}
+
+	return rules, nil
+}
+
+func validate(strs StringGenerator) (err error) {
+	merr := &multierror.Error{}
+	if strs.Length < 1 {
```

Comment:
Minimum length should be a constant.  On that topic: does it make sense to have our lower bound at 1?  Maybe 4 makes more sense?  As per benchmarking results you've shared, a significantly complicated ruleset could take a lot of time to generate a password.

Reply:
In the context of generating passwords I agree that this should be >1, however I want to keep this part of the library generic enough that we could use it for any string generation. Similar to the upper limit, I'm hesitant to make a lower limit >1.

Reply:
re: making it a constant: I don't really see the value in making it into a constant unless it's used in more than one place. Since the only place it's mentioned is here it seems excessive.